### DISCUSSION TITLE
out of heap space ?

### DISCUSSION THREAD

**Discussion by seccentral** [o3de/o3de#2077] [General]
16 GB ram not enough for compilation ? 

E:\o3de-build\External\LyShine-67f2e481\Code\CMakeFiles\LyShine.Editor.Static.dir\Unity\unity_10_cxx.cxx(19,1): fatal e
rror C1060: compiler is out of heap space [E:\o3de-build\External\LyShine-67f2e481\Code\LyShine.Editor.Static.vcxproj]
E:\o3de-cache\packages\qt-5.15.2-rev4-windows\qt\include\QtCore/qmetatype.h(1157,1): fatal error C1060: compiler is out
 of heap space [E:\o3de-build\External\PhysX-c771218d\Code\PhysX.Editor.Static.vcxproj]

**Comment by CntrlX** [comment]
i have the same problem too

**Reply by SelfishOlex** [reply]
It sounds your system ran out of memory while compiling. What are your system specs?

**Reply by seccentral** [reply]
it s the first thing in my post. read again

**Comment by rgba16f** [comment]
We do occasionally observe compiler heap space issues in visual studio. Usually these are remedied by restarting visual studio.

The error is more to do with the compile processing being limited as a 32bit process than the machine having insufficient ram.

**Reply by seccentral** [reply]
i am following the docs on o3de cli install page. i never opened vs. using cli cmake. it is 64 bit https://github.com/Kitware/CMake/releases/download/v3.20.5/cmake-3.20.5-windows-x86_64.msi

**Reply by rgba16f** [reply]
Even though you are using cmake to start the build, cmake forwards the build to the correct tools for the platform you are running on. On Windows that is primarily visual studio's command line build tools, and the visual studio compiler does occasionally run into a 32 bit process memory limitation.

**Comment by AMZN-alexpete** [comment]
I have 16GB ram on my home machine and a similar thing happened to me - but I resolved it by switched my pagefile from a slow spinny drive to my NVMe drive and it worked fine after that, though I'm sure it did some swapping.  Another suggestion I was going to try if that didn't work was to try adjusting the number of parallel processes used for compiling.    If you're still experiencing this issue can you try adjusting the number of parallel processes using the '/m:#' option?  For example, to set to 4, the build command would look something like this:

cmake --build <path/to/build/folder> --target <target> --config profile -- /m:4

Another [suggestion from Microsoft is to use the /Zm](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-errors-1/fatal-error-c1060?redirectedfrom=MSDN&view=msvc-160) option.  Hope one of these suggestions works for you.

**Comment by amznestebanpapp** [comment]
Our measurements of memory usage is around 2Gb per core. So, if you have 8 cores, you should have at least 16Gb of RAM (available).
If you don't have enough, here is a list of things to tweak:

- tweak your page file, increase and make it at least twice your amount of RAM
- reduce the amount of "project parallelization", instead of passing /m, pass something like /m:2, that will run just 2 projects at the same time. The heaviest memory consumption comes from linking big libraries, which happens once per project, therefore tweaking this setting should reduce your memory usage.
- reduce the amount of "parallel compilation within the project", add /p:CL_MPCount=4 (after the /m:2), that will run two compilations within the project at the same time.
- remove test targets from the build, pass -DLY_DISABLE_TEST_MODULES=1 when generating the project/solution. This will reduce the amount of targets, but those targets dont consume lots of memory. This is mainly to reduce compilation time.

The items above will cause slower builds.

We have been talking with some people and it seems that we will be changing the documentation to add a "recommended" spec. If your spec is bellow the recommended, we will add these tweaks to the docs.

Related: https://github.com/o3de/o3de/issues/2082
